Userland/Services/SpiceAgent/MessageHeader.cpp

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2023, Caoimhe Byrne <caoimhebyrne06@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "MessageHeader.h"
+#include <AK/Stream.h>
+#include <AK/String.h>
+
+namespace SpiceAgent {
+
+ErrorOr<MessageHeader> MessageHeader::create(Message::Type type, u32 data_size, u32 protocol_version, u64 opaque)
+{
+    return MessageHeader(protocol_version, type, opaque, data_size);
+}
+
+ErrorOr<MessageHeader> MessageHeader::read_from_stream(AK::Stream& stream)
+{
+    // The protocol version must match our agent's
+    auto protocol_version = TRY(stream.read_value<u32>());
+    if (protocol_version != AGENT_PROTOCOL) {
+        return Error::from_string_literal("Received mismatched protocol version when reading a message's header!");
+    }
+
+    // The type indicates how we should parse the message's data
+    auto type = TRY(stream.read_value<Message::Type>());
+
+    // A placeholder for messages that only pass one integer as their data
+    auto opaque = TRY(stream.read_value<u64>());
+
+    // The size of the message's data, which is after this u32
+    auto size = TRY(stream.read_value<u32>());
+
+    return MessageHeader::create(type, size, protocol_version, opaque);
+}
+
+ErrorOr<void> MessageHeader::write_to_stream(AK::Stream& stream)
+{
+    // Write all of our data to the stream
+    TRY(stream.write_value(m_protocol_version));
+    TRY(stream.write_value(m_type));
+    TRY(stream.write_value(m_opaque));
+    TRY(stream.write_value(m_data_size));
+
+    return {};
+}
```

Comment:
This looks like we could just declare `MessageHeader` as a trivially serializable object instead of reading each member in order separately.

Reply:
What do you mean by a "trivially serializable object"?

Reply:
A trivially serializable object is an object that can be constructed by simply reading/writing a span of bytes and then treating those bytes as the object. This only works if the in-memory representation matches the serialized format on disk and you don't need to run anything special while constructing/deconstructing.

Both of these seem to be the case here, so you could simply add `[[gnu::packed]]` to the `class` to eliminate any padding that the compiler may place and [declare it as trivially serializable](https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibDNS/PacketHeader.h#L101-L104). Once the latter is done, `Stream` will provide a default implementation of `write_to_stream` and `read_to_stream` so you don't have to serialize and deserialize every single member by hand.

Reply:
Question, do these implementations use the constructor?